subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/client/local/LocalReservationClient.kt

```diff
@@ -89,19 +85,54 @@ internal class LocalReservationClient(
         reservationDeployments: Collection<String>
     ) {
         if (state !is State.Reserving) {
+            // TODO: check the state on client side beforehand
             val error = RuntimeException("Unable to stop reservation job. Hasn't started yet")
             logger.critical(error.message.orEmpty())
             throw error
         }
-        (state as State.Reserving).channel.close()
+        (state as State.Reserving).devices.close()
 
         state = State.Idling
 
         logger.info("Devices released for configuration: $configurationName")
     }
 
+    private fun listenEmulators(reservation: Reservation.Data, devices: SendChannel<WorkerDevice>) {
+        // TODO: Don't use global scope. Unconfined coroutines lead to leaks
+        GlobalScope.launch {
+            logger.debug("Start listening devices for $reservation")
+            // TODO: prevent reusing the same device in different reservations
+            val reservedDevices = mutableSetOf<Serial>()
+            var emulators = localEmulators(reservation)
+
+            while (!devices.isClosedForSend) {
+                findRemainingDevices(reservation, emulators, reservedDevices)
+                    .forEach { emulator ->
+                        devices.send(emulator)
+                        reservedDevices.add(emulator.id)
+                    }
+
+                delay(TimeUnit.SECONDS.toMillis(5))
+
+                emulators = localEmulators(reservation)
+            }
+            logger.debug("Finish listening devices for $reservation")
+        }
+    }
+
+    private fun findRemainingDevices(
+        reservation: Reservation.Data,
+        devices: Set<WorkerDevice>,
+        reserved: Set<Serial>
+    ): List<WorkerDevice> {
+        val remaining = reservation.count - reserved.size
+        return devices.filterNot { device ->
+            reserved.contains(device.id)
+        }.take(remaining)
+    }
```

Comment:
I've fixed reservation size for local emulators.
In comparison to k8s, reservation state is local and stored in this client.